module/modmanager/manager.go

```diff
@@ -121,6 +135,83 @@ func (mgr *Manager) Add(ctx context.Context, cfg config.Module) error {
 	return nil
 }
 
+// Reconfigure reconfigures an existing resource module and returns the names of
+// now orphaned resources.
+func (mgr *Manager) Reconfigure(ctx context.Context, cfg config.Module) ([]resource.Name, error) {
+	mod, exists := mgr.modules[cfg.Name]
+	if !exists {
+		return nil, errors.Errorf("cannot reconfigure module %s as it does not exist", cfg.Name)
+	}
+	handledResources := mod.resources
+
+	if err := mgr.remove(mod); err != nil {
+		return nil, err
+	}
+
+	if err := mgr.add(ctx, cfg); err != nil {
+		return nil, err
+	}
+
+	// add old module process' resources to new module; warn if new module cannot
+	// handle old resource and consider that resource orphaned.
+	var orphanedResourceNames []resource.Name
+	for name, res := range handledResources {
+		if _, err := mgr.AddResource(ctx, res.cfg, res.deps); err != nil {
+			mgr.logger.Warnf("error while re-adding resource %s to module %s: %v",
+				name, cfg.Name, err)
+			orphanedResourceNames = append(orphanedResourceNames, name)
+		}
+	}
+	return orphanedResourceNames, nil
+}
+
+// Remove removes and stops an existing resource module and returns the names of
+// now orphaned resources.
+func (mgr *Manager) Remove(modName string) ([]resource.Name, error) {
+	mod, exists := mgr.modules[modName]
+	if !exists {
+		return nil, errors.Errorf("cannot remove module %s as it does not exist", modName)
+	}
+	handledResources := mod.resources
+
+	if err := mgr.remove(mod); err != nil {
+		return nil, err
+	}
+
+	var orphanedResourceNames []resource.Name
+	for name := range handledResources {
+		orphanedResourceNames = append(orphanedResourceNames, name)
+	}
+	return orphanedResourceNames, nil
+}
+
+func (mgr *Manager) remove(mod *module) error {
+	mgr.mu.Lock()
+	defer mgr.mu.Unlock()
+
+	if err := mod.stopProcess(); err != nil {
+		return errors.WithMessage(err, "error while stopping module "+mod.name)
+	}
+
+	if mod.conn != nil {
+		if err := mod.conn.Close(); err != nil {
```

Comment:
Previously, I was not closing the gRPC connection to the module when reconfiguring. I think we actually have to close this connection, as when we `dial` again in `add`, the old `conn` is leaked.

Reply:
So if that IS actually happening, this probably isn't the right (or at least sufficient) fix. The actual resources from a module use the same connection as the module itself, so by closing it, it seems like you're now disconnecting all those resources, and above in the reconfigure, you're running an "AddResource()" loop, but ignoring the newly returned clients (using the new module's new connection) so this seems like it would orphan a lot of resource clients (e.g. resource manager nodes) with broken connections.

Reply:
Either those existing resources (that may already be in use by client code, or as dependencies of other resources) will reconnect to the new version of the module, or this is going to get a LOT more complicated, and new clients with new connections (as returned by mgr.AddResource()) will have to somehow be swapped out for the previous client instances.